ext/fg/js/frontend.js

```diff
@@ -262,7 +371,97 @@ class Frontend {
 
         this.textSourceLast = null;
     }
+
+    getPrimaryTouch(touchList) {
+        return touchList.length > 0 ? touchList[0] : null;
+    }
+
+    getIndexOfTouch(touchList, identifier) {
+        for (let i = 0, ii = touchList.length; i < ii; ++i) {
+            let t = touchList[i];
+            if (t.identifier === identifier) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    excludeTouches(touchList, excludeTouchList) {
+        const result = [];
+        for (let i = 0, ii = touchList.length; i < ii; ++i) {
+            let r = result[i];
+            if (this.getIndexOfTouch(excludeTouchList, r.identifier) < 0) {
+                result.push(r);
+            }
+        }
+        return result;
+    }
+
+    setPrimaryTouch(touch) {
+        if (touch === null) {
+            this.primaryTouchIdentifier = null;
+            this.contextMenuPreviousRange = null;
+            this.contextMenuChecking = false;
+            this.scrollPrevent = false;
+            this.setContextMenuPrevent(false, true);
+            this.setMouseDownPrevent(false, true);
+            this.setClickPrevent(false);
+        }
+        else {
+            this.primaryTouchIdentifier = touch.identifier;
+            this.contextMenuPreviousRange = this.textSourceLast ? this.textSourceLast.clone() : null;
+            this.contextMenuChecking = true;
+            this.scrollPrevent = false;
+            this.setContextMenuPrevent(false, false);
+            this.setMouseDownPrevent(false, false);
+            this.setClickPrevent(false);
+
+            this.searchFromTouch(touch.clientX, touch.clientY, Frontend.SearchType.TouchStart);
+        }
+    }
+
+    setContextMenuPrevent(value, delay) {
+        if (!delay) {
+            this.contextMenuPrevent = value;
+        }
+    }
+
+    setMouseDownPrevent(value, delay) {
+        if (!delay) {
+            this.mouseDownPrevent = value;
+        }
+    }
+
+    setClickPrevent(value) {
+        this.clickPrevent = value;
+    }
+
+    searchFromTouch(x, y, type) {
+        this.popupTimerClear();
+
+        if (!this.options.general.enable || this.pendingLookup) {
+            return;
+        }
+
+        const pos = {x: x, y: y};
```

Comment:
This can actually just be `{x, y}`, and at the point where it's that short, you don't even have to have separate constant for it : )

Reply:
Not sure why I overcomplicated that. Fixed.